packages/babel-plugin-transform-async-to-generator/test/fixtures/async-to-generator/async-complex-params/output.js

```diff
@@ -0,0 +1,49 @@
+new class {
+  foo(_x, _x2) {
+    return babelHelpers.asyncToGenerator(function* (a, {
+      b
+    }) {}).apply(this, arguments);
+  }
+}().foo().catch(e => {
+  console.log("caught");
+});
+new class {
+  foo(_x3, _x4) {
+    return babelHelpers.asyncToGenerator(function* (a, {
+      b
+    }) {
+      console.log(a, b);
+    }).apply(this, arguments);
+  }
+}().foo(1, {
+  b: 2
+});
+new class {
+  foo(_x5, _x6) {
+    var _arguments = arguments,
+      _this = this;
+    return babelHelpers.asyncToGenerator(function* (a, {
+      b
+    }) {
+      console.log(_this, _arguments);
+    }).apply(this, arguments);
+  }
+}().foo(1, {
```

Comment:
Will this work?
```
  foo(_x5, _x6) {
    var _arguments = arguments,
      _this = this;
    return babelHelpers.asyncToGenerator(function* (a, {
      b
    }) {
      console.log(_this, _arguments);
    })(_x5, _x6);
  }
```

Reply:
This approach seems feasible, but I encountered this scenario during testing:

```JavaScript
"use strict";
class Foo {
    async foo(a, b = 2, c) {
        return a + b + c;
    }
}
```

Our parameter copying logic stops at `b = 2` because the `length` property of the `foo` method in Foo instances reports 1. This causes parameters to be lost during the passing process.

Reply:
Alternatively, here's a more straightforward example:

```JavaScript
class Foo {
    async foo(a, b = 2) {
        return a + b;
    }
}
```

Reply:
@liuxingbaoyu It might be worth considering avoiding the hoisting of this and arguments in this scenario, as they have already been properly passed. 
That said, this adjustment could likely be deferred to a follow-up PR to keep the current changeset focused. What do you think?

Reply:
Yes, I have a PR planned to remove the hoisting in functions, we can also consider whether to do the same for methods.
https://github.com/babel/babel/pull/15992